<div align='center' >
<p><font size='70'><strong>PanGraphRNA User Manual</strong></font></p>
<font size='100'>(version 1.0)</font>
</div>

- PanGraphRNA is an efficient, flexible and web-based Galaxy platform that can be easily used to construct graph pangenomes from genetic variations at individual, subpopulation, and population levels. It can assist researchers to select appropriate graph pangenomes using various performance metrics for both real and simulation experiments. 
- Currently, PanGraphRNA is composed of four functional modules: **Graph Pangenome Preparation Module, Construction Module, Evaluation Module, and Application Moudule**.
- PanGraphRNA was powered with an advanced  packaging technology, which enables compatibility and portability.
- PanGraphRNA project is hosted on [https://github.com/cma2015/PanGraphRNA](https://github.com/cma2015/PanGraphRNA)
- PanGraphRNA docker image is available at [https://hub.docker.com/r/malab/pangraphrna](https://hub.docker.com/r/malab/pangraphrna)

## Graph Pangenome Evaluation Module

This module assesses the performance of graph pangenomes against the SLR strategy with six read mapping-relevant measurements: _Unique alignment rate_, _Multiple alignment rate_, _Error mapping rate_, _F1 score_, _Recall_ and _Precision_.

| **Tools**                       | **Description**                                              | **Input**                                       | **Output**                                        | **Time (test data)**         | **Reference**                                                |
| ------------------------------- | ------------------------------------------------------------ | ----------------------------------------------- | ------------------------------------------------- | ---------------------------- | ------------------------------------------------------------ |
| **Basic Alignment Statistics**              | Obtain the bascic alignment information among multiple read-genome alignment results | HISAT2 alignment report in TXT format and graph pangenome files                 | Basic alignment statistics matrix in TXT format                    | ~1 mins | / |
| **Mapping Error Measurement**               | Perform RNA-seq reads simulation and measure mapping errors | Ground truth files and graph pangenome files                 | Measurement of mapping errors in CSV format                    | ~10 mins | <a href="http://flux.sammeth.net/" target="_blank">Flux</a> and <a href="https://github.com/DaehwanKimLab/hisat2" target="_blank">HISAT2</a> |
| **F1 Score Calculation** | Perform RNA-seq reads simulation and calculate F1 score | Graph pangenome files | Calculation of F1 score in CSV format | ~10 mins                      | <a href="http://flux.sammeth.net/" target="_blank">Flux</a> and <a href="https://github.com/DaehwanKimLab/hisat2" target="_blank">HISAT2</a> |

## Basic Alignment Statistics

This function is designed to count uniquely mapped reads, multiply mapped reads and calculate alignment rates among multiple read-genome alignment results.

#### Input

-   **Input HISAT2 alignment report file:** Input alignment summary files generated by HISAT2 in TXT format
-   **Input reference genome file:** Input reference genome file for primary path of graph pangenome in FASTA format
  
#### Output

-   **Alignment information matrix in TXT format**

![[Figure3_1.jpg]]

## Mapping Error Measurement

**Mapping error measurement** is designed to identify incorrectly mapped reads and record their error types in RNA-seq read-genome alignments. This function also enables **Flux** features (Griebel, T., _et al_., 2012) for generating simulated reads and performs read-genome alignment using **HISAT2** (Kim, D., _et al_., 2019). For more details, see [https://bio.tools/the_flux_simulator](https://bio.tools/the_flux_simulator) and [https://daehwankimlab.github.io/hisat2/manual](https://daehwankimlab.github.io/hisat2/manual) for details.

**Measuring mapping errors**

  The alignment results for different reference genomes were compared with the ground truth, and the five error types were defined according to the read alignment type and position.

  - Five **mapping error** types:

| Mapping error type| Groundtruth alignment result (read type) | Alignment result (read type) |
| ------------------------------- | ------------------------------------------------------------ | ----------------------------------------------- |
| Mapped to UnMapped|Mapped read| Unmapped read|
| UnMapped to Mapped|Unmapped read| Mapped read|
| Unique to MultiMapper|Uniquely mapped read|Multiply mapped read|
| MultiMapper to Unique|Multiply mapped read|Uniquely mapped read|
| Different Mapping Loci|Mapped to A region|Mapped to B region|

#### Input

-   **Ground Truth Setting**
-   **Input ground truth genome file:** Input ground truth reference genome in FASTA format
-   **Input chain file:** Input chain file of ground truth genome (GroundTruth to ReferenceGenome) in CHAIN format

-   **Graph Pangenome Information**
-   **Input reference genome file:** Input reference genome file that has been used for graph pangenome construction in FASTA format
-   **Input VCF file:** Input vcf file that has been used for graph pangenome construction in VCF format
-   **Input exon annotation file:** Input exon annotation file (GTF file) of reference genome file used for constructing the primary path of graph pangenome in GTF format

-   **RNA-Seq Reads Simulation**
-   **Input VCF file:** Input VCF file containing variation information for pseudo genome construction in VCF format
  
#### Parameters

-   **Generation of simulated RNA-seq datasets:** Input number of how many sets of simulated reads you need to generate (Default: 15)
-   **Simulated RNA-seq read number:** Input simreads number to generate for each dataset (Default: 1000000)
-   **Threads:** The number of threads used for parallel computation (Default: 10)

#### Output

-   **Measurement of mapping errors in CSV format**

![[Figure3_2.jpg]]

## F1 Score Calculation

**Mapping error measurement** is designed to calculate **Recall**, **Precision** and **F1 score** based on RNA-seq read-genome alignments. This function also enables **Flux** features (Griebel, T., _et al_., 2012) for generating simulated reads and performs read-genome alignment using **HISAT2** (Kim, D., _et al_., 2019). For more details, see [https://bio.tools/the_flux_simulator](https://bio.tools/the_flux_simulator) and [https://daehwankimlab.github.io/hisat2/manual](https://daehwankimlab.github.io/hisat2/manual) for details.

**Measuring F1 score**

-   **Recall** = 100 X _Num.correct_ / _Num.reads_
-   **Precision** = 100 X _Num.correct_ / _Num.unique_
-   **F1 score** = ( 2 X _Recall_ X _Precision_ ) / ( _Recall_ + _Precision_ )
-   _Num.reads_ represent the total number of reads in the simulated RNA-seq data; _Num.correct_ represent the number of correctly aligned reads, matching the coordinates of the simulated read generation position; and _Num.unique_ represent the number of uniquely mapped reads, with the uniquely mapped tag in the HISAT2 alignment result being "NH:i:1".

#### Input

-   **Graph Pangenome Information**
-   **Input FASTA files:** Input the reference genome file that have been used for graph pangenome construction in FASTA format
-   **Input VCF files:** Input the vcf file that have been used for graph pangenome construction in VCF format
-   **Input GTF files:** Input the exon annotion file of reference genome that have been used for graph pangenome construction in GTF format

-   **RNA-Seq Reads Simulation**
-   **Input VCF files:** Input the vcf file containing the accession name for pseudo genome construction in VCF format

#### Parameters

-   **Generation of simulated RNA-seq datasets:** Input number of how many sets of simulated reads you need to generate (Default: 15)
-   **Simulated RNA-seq read number:** Input simreads number to generate for each dataset (Default: 1000000)
-   **Threads:** The number of threads used for parallel computation (Default: 10)

#### Output

-   **Calculation of F1 score in CSV format**

![[Figure3_3.jpg]]